pkg/scheduler/framework/plugins/podtopologyspread/plugin.go

```diff
@@ -141,9 +149,119 @@ func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEventWithHint
+	if err != nil {
+		pl.logger.Error(err, "unexpected objects in isSchedulableAfterPodChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.Queue, err
+	}
+
+	constraints, err := pl.getConstraints(pod)
+	if err != nil {
+		pl.logger.Error(err, "get constraints from pod failed:", "pod", klog.KObj(pod))
+		return framework.Queue, err
+	}
+
+	if modifiedPod != nil && originalPod != nil && !reflect.DeepEqual(modifiedPod.Labels, originalPod.Labels) {
+		for _, c := range constraints {
+			if c.Selector.Matches(labels.Set(originalPod.Labels)) || c.Selector.Matches(labels.Set(modifiedPod.Labels)) {
+				return framework.Queue, nil
+			}
+		}
+		return framework.QueueSkip, nil
 	}
+
+	if originalPod != nil {
+		for _, c := range constraints {
+			if c.Selector.Matches(labels.Set(originalPod.Labels)) {
+				return framework.Queue, nil
+			}
+		}
+	}
+
+	if modifiedPod != nil {
+		for _, c := range constraints {
+			if c.Selector.Matches(labels.Set(modifiedPod.Labels)) {
+				return framework.Queue, nil
+			}
+		}
+	}
+
+	return framework.QueueSkip, nil
+}
+
+func (pl *PodTopologySpread) getConstraints(pod *v1.Pod) ([]topologySpreadConstraint, error) {
+	var constraints []topologySpreadConstraint
+	var err error
+	if len(pod.Spec.TopologySpreadConstraints) > 0 {
+		// We have feature gating in APIServer to strip the spec
+		// so don't need to re-check feature gate, just check length of Constraints.
+		constraints, err = pl.filterTopologySpreadConstraints(
+			pod.Spec.TopologySpreadConstraints,
+			pod.Labels,
+			v1.DoNotSchedule,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("obtaining pod's hard topology spread constraints: %w", err)
+		}
+	} else {
+		constraints, err = pl.buildDefaultConstraints(pod, v1.DoNotSchedule)
+		if err != nil {
+			return nil, fmt.Errorf("setting default hard topology spread constraints: %w", err)
+		}
+	}
+	return constraints, nil
+}
+
+func (pl *PodTopologySpread) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		pl.logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.Queue, err
+	}
+
+	constraints, err := pl.getConstraints(pod)
+	if err != nil {
+		pl.logger.Error(err, "get constraints from pod failed:", "pod", klog.KObj(pod))
+		return framework.Queue, err
+	}
+
+	// framework.UpdateNodeLabel: return Queue when topologyKey is added/removed in node's labels, else return QueueSkip.
+	if originalNode != nil && modifiedNode != nil {
+		pl.logger.V(4).Info("node was updated", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		if nodeLabelsMatchSpreadConstraints(originalNode.Labels, constraints) {
+			return framework.Queue, nil
+		}
+		if nodeLabelsMatchSpreadConstraints(modifiedNode.Labels, constraints) {
+			return framework.Queue, nil
+		}
```

Comment:
It isn't consistent with L236 comment since it always return Queue if `originalNode.Labels` matches with constraints.

Reply:
Do you mean that it's not necessary to return Queue when both `originalNode.Labels` and `modifiedNode.Labels` match with constraints but label value not changed?

Reply:
> // framework.UpdateNodeLabel: return Queue when topologyKey is added/changed/removed in node's labels, else return QueueSkip.

So, this ↑ comment is correct, but the current commit doesn't implement as this comment.

It'd return Queue even when the label is not changed. 
For example, topologyKey is "zone", then here we'd like to return Queue:
- when node doesn't have "zone" label at first, and "zone" label is newly added to the node.
- when "zone" label is updated from zone-a to zone-c
- when "zone" label is deleted.

Reply:
> It'd return Queue even when the label is not changed.

it will enter `isSchedulableAfterNodeChange` only when `UpdateNodeLabel/Add/Delete` event. So I don't think there is a scenario which orignalNode and modifiedNode have the *same* labels.

Reply:
It'll enter when non-related labels are updated, won't it?